dragnet/_readability.cc

```diff
@@ -0,0 +1,124 @@
+
+#include <vector>
+#include <string>
+#include <unordered_map>
+
+// NOTE: features is the return vector.  we assume it is preinitialized
+// to length(nblocks) with 0.0
+
+const std::string DIV = "div";
+const std::string P = "p";
+
+void _readability_features(
+    std::vector<uint32_t>& block_text_len,
+    std::vector<std::vector<std::pair<uint32_t, int> > >& block_readability_class_weights,
+    std::vector<std::vector<uint32_t> >& block_ancestors,
+    std::vector<std::string>& block_start_tag,
+    std::vector<double>& block_link_density,
+    int& nblocks,
+    double* features)
+{
+    std::unordered_map<uint32_t, double> scores;
+    // link density in subtrees.
+    // first entry is total text length * link density,
+    // second is total text length
+    std::unordered_map<uint32_t, std::pair<double, double> > ld;
+    // only tag_ids that have <p> or <div> as children are valid root nodes
+    std::unordered_map<uint32_t, bool> valid_nodes;
+
+    std::vector<std::pair<uint32_t, int> >::const_iterator it_cw;
```

Comment:
It's generally best to declare loop variables directly inside their corresponding loop unless you need them beyond the scope of the loop. Reducing variable scope decreases the amount of mental state that needs to be maintained by the readers of the code.